src/index.js

```diff
@@ -1,169 +1,323 @@
+	 *
+	 * @param {PopStateEvent|boolean} popstate
+	 * @returns void
+	 */
+	doScrollingBetweenPages = (popstate) => {
+		const swup = this.swup;
+
+		// Bail early on popstate and inactive `animateHistoryBrowsing`
+		if (popstate && !swup.options.animateHistoryBrowsing) {
+			return;
+		}
+
+		// Try scrolling to a given anchor
+		if (this.maybeScrollToAnchor(swup.scrollToElement, 'betweenPages')) {
+			swup.scrollToElement = null;
+			return;
+		}
+
+		// Finally, scroll to either the stored scroll position or to the very top of the page
+		const scrollPositions = this.getStoredScrollPositions(getCurrentUrl());
+		const top = (scrollPositions.window && scrollPositions.window.top) || 0;
+		swup.scrollTo(top, this.shouldAnimate('betweenPages'));
+	};
+
+	/**
+	 * Stores the current scroll positions for the URL we just came from
+	 */
+	onWillReplaceContent = () => {
+		this.storeScrollPositions(this.previousUrl);
+		this.previousUrl = getCurrentUrl();
+	};
+
+	/**
+	 * Deletes the scroll positions for the URL a link is pointing to,
+	 * if shouldRestoreScrollPosition doesn't evaluate to true
+	 *
+	 * @param {event} e
+	 * @returns void
+	 */
+	onClickLink = (e) => {
+		if (!this.options.shouldRestoreScrollPosition(e.delegateTarget)) {
+			return;
+		}
+		const url = new Link(e.delegateTarget).getAddress();
+		this.deleteStoredScrollPositions(url);
+	};
+
+	/**
+	 * Stores the scroll positions for the current URL
+	 * @param {string} url
+	 */
+	storeScrollPositions(url) {
+		// create an object to store the scroll positions
+		const storeEntry = {
+			window: { top: window.scrollY, left: window.scrollX },
+			containers: []
+		};
+		// fill up the object with scroll positions for each matching element
+		const $containers = document.querySelectorAll(this.options.scrollContainers);
+		$containers.forEach((el) =>
+			storeEntry.containers.push({
+				top: el.scrollTop,
+				left: el.scrollLeft
+			})
+		);
+		// put the object into the store
+		this.scrollPositionsStore[url] = storeEntry;
+	}
+
+	/**
+	 * Deletes stored scroll positions for a given URL
+	 * @param {string} url
+	 */
+	deleteStoredScrollPositions(url) {
+		delete this.scrollPositionsStore[url];
+		this.scrollPositionsStore[url] = null;
+	}
+
+	/**
+	 * Get the stored scroll positions for a given URL from the cache
+	 * @returns {object}
+	 */
+	getStoredScrollPositions(url) {
+		return this.scrollPositionsStore[url] || {};
+	}
+
+	/**
+	 * Restore the scroll positions for all matching scrollContainers
+	 * @returns void
+	 */
+	restoreScrollContainers(popstate) {
+		// get the stored scroll positions from the cache
+		const scrollPositions = this.getStoredScrollPositions(getCurrentUrl());
+		if (scrollPositions.containers == null) {
+			return;
+		}
+
+		// cycle through all containers on the current page and restore their scroll positions, if appropriate
+		const $containers = document.querySelectorAll(this.options.scrollContainers);
+		$containers.forEach((el, index) => {
```

Comment:
For consistency and simplicity, `queryAll(this.options.scrollContainers).forEach()`